lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+			case <-returnedChan:
+			}
+
+			return
+		}
+
+		close(regSentChan)
+	}()
+
+	select {
+	case err := <-errChan:
+		return fmt.Errorf("error sending registration complete "+
+			"message to remote signer: %w", err)
+
+	case <-streamCtx.Done():
+		return streamCtx.Err()
+
+	case <-r.quit:
+		return ErrShuttingDown
+
+	case <-time.After(r.requestTimeout):
+		return errors.New("watch-only node handshake timeout")
+
+	case <-regSentChan:
+	}
+
+	// After the registration message has been sent, the signer node will
+	// respond with a message indicating that it has accepted the signer
+	// registration request if the registration was successful.
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		msg, err := r.stream.Recv()
+		if err != nil {
+			select {
+			case errChan <- err:
+			case <-returnedChan:
+			}
+
+			return
+		}
+
+		// Verify that the request ID of the response is the same as the
+		// request ID of the registration message.
+		if msg.GetRequestId() != handshakeRequestID {
+			err = fmt.Errorf("initial response request id must "+
+				"be %d, but is: %d", handshakeRequestID,
+				msg.GetRequestId())
+
+			select {
+			case errChan <- err:
+			case <-returnedChan:
+			}
+
+			return
+		}
+
+		// Check the type of the response message.
+		switch reqType := msg.GetSignRequestType().(type) {
+		case *walletrpc.SignCoordinatorRequest_RegistrationResponse:
+
+			switch rType := reqType.RegistrationResponse.
+				GetRegistrationResponseType().(type) {
+			// The registration was successful.
+			case *completeType:
+				select {
+				case completeChan <- rType.RegistrationComplete:
+				case <-returnedChan:
+				}
+
+			// An error occurred during the registration process.
+			case *walletrpc.RegistrationResponse_RegistrationError:
+				err := fmt.Errorf("registration error: %s",
+					rType.RegistrationError)
+
+				select {
+				case errChan <- err:
+				case <-returnedChan:
+				}
+			}
+
+			return
+
+		default:
+			err := fmt.Errorf("expected registration response, "+
+				"but got: %T", reqType)
+
+			select {
+			case errChan <- err:
+			case <-returnedChan:
+			}
+
+			return
+		}
+	}()
+
+	// Wait for the watch-only node to respond that it has accepted the
+	// signer has registered.
+	select {
+	case <-completeChan:
+		// TODO(viktor): This should verify that the signature in the
+		// complete message is valid.
+
+	case err := <-errChan:
+		return fmt.Errorf("watch-only node handshake error: %w", err)
+
+	case <-r.quit:
+		return ErrShuttingDown
+
+	case <-streamCtx.Done():
+		return streamCtx.Err()
+
+	case <-time.After(r.requestTimeout):
+		return errors.New("watch-only node handshake timeout")
+	}
+
+	return nil
+}
+
+// processSignRequests processes and responds to the sign requests that are
+// sent over the stream. The function will continuously run until the remote
+// signer client is either stopped or the stream errors.
+func (r *OutboundClient) processSignRequests(streamCtx context.Context) error {
```

Comment:
```
func (r *OutboundClient) processSignRequests(ctx context.Context,
	stream StreamClient) error {
	var (
		reqChan = make(chan *walletrpc.SignCoordinatorRequest)
		errChan = make(chan error)
	)
	defer close(reqChan)
	defer close(errChan)

	// We run the receive loop in a goroutine to ensure we can stop if the
	// remote signer client is shutting down (i.e. the quit channel is
	// closed). Closing the quit channel will make the processSignRequests
	// function return, which will cancel the stream context, which in turn
	// will stop the receive goroutine.
	r.wg.Add(1)
	go func() {
		defer r.wg.Done()

		for {
			req, err := stream.Recv()
			if err != nil {
				wrappedErr := fmt.Errorf("error receiving "+
					"request from watch-only node: %w", err)

				errChan <- wrappedErr

				return
			}

			reqChan <- req
		}
	}()

	for {
		log.Tracef("Waiting for a request from the watch-only node")

		select {
		case req := <-reqChan:
			// Process the received request.
			err := r.handleRequest(ctx, req, stream)
			if err != nil {
				return err
			}

		case <-r.quit:
			return ErrShuttingDown

		case err := <-errChan:
			return err
		}
	}
}
```

Reply:
Thanks :pray:! I implemented parts of this, but also didn't change some of the suggestions. Leaving comments below on stuff not changed:

```
func (r *OutboundClient) processSignRequests(ctx context.Context,
	stream StreamClient) error {
	var (
		reqChan = make(chan *walletrpc.SignCoordinatorRequest)
		errChan = make(chan error)
	)
------
	We intentionally don't want to close the channels here, as there would be no guarantee that the goroutine below has finished it's execution before these channels are closed. That would to a panic. So instead we let the garbage collector remove these once this function + goroutine has finished executing. I added a comment in the code to clarify this.
-------
	defer close(reqChan)
	defer close(errChan)

	// We run the receive loop in a goroutine to ensure we can stop if the
	// remote signer client is shutting down (i.e. the quit channel is
	// closed). Closing the quit channel will make the processSignRequests
	// function return, which will cancel the stream context, which in turn
	// will stop the receive goroutine.
	r.wg.Add(1)
	go func() {
		defer r.wg.Done()

		for {
			req, err := stream.Recv()
			if err != nil {
				wrappedErr := fmt.Errorf("error receiving "+
					"request from watch-only node: %w", err)

------------
				We must check if the context has been canceled here, as that means that the outer function has returned and closed the stream.
				If we do not do so, this will be stuck on the sending over the channel,
				and therefore never finishing the execution of the goroutine.
				That would block the shutting down of lnd.
				Removing the r.quit should be safe here though, so I went ahead and did so :).
---------
				errChan <- wrappedErr

				return
			}

			reqChan <- req
		}
	}()

	for {
		log.Tracef("Waiting for a request from the watch-only node")

------------
Hmmmm not sure if we'd like to remove the ctx.Done() check here now that this an external context that's been passed in through lnd? I guess that you could argue that that context will never be cancelled before lnd is shutdown, which would trigger r.quit anyways. But that feels like an assumption that is quite unnecessary to make, and may lead to bugs? 
------------
		select {
		case req := <-reqChan:
			// Process the received request.
			err := r.handleRequest(ctx, req, stream)
			if err != nil {
				return err
			}

		case <-r.quit:
			return ErrShuttingDown

		case err := <-errChan:
			return err
		}
	}
}
```

Reply:
Gotcha :)